File: packages/ts/server/src/enrichers/entitlement-role-enricher.ts
import type { ObjectStore } from "../storage/base.js"
import type { RoleEnricher, AuthResult } from "../router/route-builder.js"

export interface EntitlementRoleEnricherOptions {
  /** The ObjectStore to read entitlement documents from. */
  store: ObjectStore
  /**
   * Storage path template for the per-user entitlement document.
   * `{identity}` is replaced with the authenticated user's identity at runtime.
   * Defaults to `"users/{identity}/entitlements"`.
   */
  path?: string
  /**
   * Top-level field in the entitlement document data that holds the list of feature slugs.
   * Defaults to `"features"`.
   */
  field?: string
  /**
   * Prefix applied to each feature slug when constructing the role string.
   * A slug `"premium-package-1"` with prefix `"entitlement"` yields role
   * `"entitlement:premium-package-1"`.
   * Defaults to `"entitlement"`.
   *
   * Change this if your role namespace already uses `"entitlement:"` for something else.
   */
  rolePrefix?: string
  /**
   * How long (in milliseconds) to cache entitlement lookups per user.
   * Set to 0 to disable caching. Defaults to 60 000 (1 minute).
   *
   * Note: each active user adds one cache entry; for very high user counts
   * set this low or disable and use an external cache layer.
   */
  cacheTtlMs?: number
}

interface CacheEntry {
  features: Set<string>
  expiresAt: number
}

/**
 * Creates a `RoleEnricher` that grants roles based on a per-user entitlement document
 * stored in the ObjectStore.
 *
 * The entitlement document must be a standard Starfish JSON document whose `data`
 * field contains a string array under `field` (default `"features"`):
 *
 * ```json
 * { "features": ["premium-package-1", "paid-cloud-sync"] }
 * ```
 *
 * Each slug is translated to a role string: `"${rolePrefix}:${slug}"` (default prefix:
 * `"entitlement"`). Collections gate access using these roles in `readRoles`/`writeRoles`:
 *
 * ```ts
 * { readRoles: ["entitlement:premium-package-1"], writeRoles: ["admin"] }
 * ```
 *
 * Recommended entitlement collection config:
 * ```ts
 * { name: "entitlements", storagePath: "users/{identity}/entitlements",
 *   readRoles: ["self"], writeRoles: ["admin"],
 *   encryption: "none", maxBodyBytes: 4096, allowedMimeTypes: ["application/json"] }
 * ```
 *
 * Usage:
 * ```ts
 * const entitlementEnricher = createEntitlementRoleEnricher({ store })
 *
 * const router = createSyncRouter({
 *   store, config, roleResolver,
 *   roleEnricher: entitlementEnricher,
 * })
 * ```
 *
 * When combined with a group enricher, use `composeEnrichers`:
 * ```ts
 * import { composeEnrichers } from "@drakkar.software/starfish-server"
 * roleEnricher: composeEnrichers(groupEnricher, entitlementEnricher)
 * ```
 *
 * Note: the enricher fires for every collection request, including the entitlement
 * collection itself. For that collection `readRoles: ["self"]` already passes before
 * enrichment — the extra entitlement roles are ignored for that request.
 */
export function createEntitlementRoleEnricher(opts: EntitlementRoleEnricherOptions): RoleEnricher {
  const {
    store,
    path = "users/{identity}/entitlements",
    field = "features",
    rolePrefix = "entitlement",
    cacheTtlMs = 60_000,
  } = opts

  const cache = new Map<string, CacheEntry>()

  async function resolveFeatures(identity: string): Promise<Set<string>> {
    const now = Date.now()

    if (cacheTtlMs > 0) {
      const cached = cache.get(identity)
      if (cached && cached.expiresAt > now) {
        return cached.features
      }
    }

    const key = path.replace("{identity}", identity)
    const raw = await store.getString(key)

    let features = new Set<string>()
    if (raw != null) {
      try {
        // StoredDocument format: { v: 1, data: { features: [...] }, timestamps: {...}, hash: "..." }
        const doc = JSON.parse(raw) as { data?: Record<string, unknown> }
        const list = doc.data?.[field]
        if (Array.isArray(list)) {
          features = new Set(list.filter((s): s is string => typeof s === "string"))
        }
      } catch {
        // Corrupt document — treat as no entitlements
      }
    }

    if (cacheTtlMs > 0) {
      cache.set(identity, { features, expiresAt: now + cacheTtlMs })
    }

    return features
  }

  return async function entitlementRoleEnricher(
    auth: AuthResult,
    _params: Record<string, string>,
  ): Promise<string[]> {
    const features = await resolveFeatures(auth.identity)
    return Array.from(features, (slug) => `${rolePrefix}:${slug}`)
  }
}

File: packages/python/entitlements/starfish_entitlements/entitlement_role_enricher.py
"""Entitlement-based role enricher for Starfish."""


import json
import logging
import time
from dataclasses import dataclass, field

from starfish_server.storage.base import AbstractObjectStore
from starfish_server.router.route_builder import AuthResult, RoleEnricher


@dataclass
class EntitlementRoleEnricherOptions:
    """Options for :func:`create_entitlement_role_enricher`."""

    store: AbstractObjectStore
    """The ObjectStore to read entitlement documents from."""

    path: str = "users/{identity}/entitlements"
    """Storage path template for the per-user entitlement document.

    ``{identity}`` is replaced with the authenticated user's identity at runtime.
    Defaults to ``"users/{identity}/entitlements"``.
    """

    field: str = "features"
    """Top-level field in the entitlement document data holding the list of
    feature slugs.  Defaults to ``"features"``."""

    role_prefix: str = "entitlement"
    """Prefix applied to each feature slug when constructing the role string.

    A slug ``"premium-package-1"`` with prefix ``"entitlement"`` yields role
    ``"entitlement:premium-package-1"``.  Defaults to ``"entitlement"``.

    Change this if your role namespace already uses ``"entitlement:"`` for
    something else.
    """

    cache_ttl_ms: int = 60_000
    """How long (in milliseconds) to cache entitlement lookups per user.
    Set to 0 to disable caching.  Defaults to 60 000 (1 minute)."""


@dataclass
class _CacheEntry:
    features: frozenset[str]
    expires_at: float


def create_entitlement_role_enricher(opts: EntitlementRoleEnricherOptions) -> RoleEnricher:
    """Return a :class:`RoleEnricher` that grants roles from a per-user entitlement document.

    The entitlement document must be a standard Starfish JSON document whose
    ``data`` field contains a string list under ``opts.field`` (default ``"features"``):

    .. code-block:: json

        { "features": ["premium-package-1", "paid-cloud-sync"] }

    Each slug is translated to a role: ``f"{opts.role_prefix}:{slug}"``.  Collections
    gate access via ``read_roles`` / ``write_roles``:

    .. code-block:: python

        CollectionConfig(
            name="premium-data",
            read_roles=["entitlement:premium-package-1"],
            ...
        )

    Recommended entitlement collection config::

        CollectionConfig(
            name="entitlements",
            storage_path="users/{identity}/entitlements",
            read_roles=["self"],
            write_roles=["admin"],
            encryption="none",
            max_body_bytes=4096,
        )

    Usage::

        enricher = create_entitlement_role_enricher(EntitlementRoleEnricherOptions(store=store))

        router = create_sync_router(SyncRouterOptions(
            store=store,
            config=config,
            role_resolver=role_resolver,
            role_enricher=enricher,
        ))

    When combined with another enricher, use :func:`compose_enrichers`::

        from starfish_server import compose_enrichers
        role_enricher=compose_enrichers(group_enricher, entitlement_enricher)
    """
    cache: dict[str, _CacheEntry] = {}

    async def resolve_features(identity: str) -> frozenset[str]:
        now = time.monotonic() * 1000

        if opts.cache_ttl_ms > 0:
            entry = cache.get(identity)
            if entry is not None and entry.expires_at > now:
                return entry.features

        key = opts.path.replace("{identity}", identity)
        raw = await opts.store.get_string(key)

        features: frozenset[str] = frozenset()
        if raw is not None:
            try:
                # StoredDocument format: { "v": 1, "data": { "features": [...] }, ... }
                doc = json.loads(raw)
                feature_list = doc.get("data", {}).get(opts.field)
                if isinstance(feature_list, list):
                    features = frozenset(s for s in feature_list if isinstance(s, str))
            except (json.JSONDecodeError, AttributeError) as exc:
                logging.getLogger(__name__).error(
                    "entitlement-enricher: corrupt entitlement document at %r: %s", key, exc
                )
                # Corrupt document — treat as no entitlements, but do NOT cache
                # the empty result (matching the TS enricher): a transient
                # corruption must not deny entitlement roles for the whole TTL
                # after the document is repaired.
                return features

        if opts.cache_ttl_ms > 0:
            cache[identity] = _CacheEntry(
                features=features,
                expires_at=now + opts.cache_ttl_ms,
            )

        return features

    async def entitlement_role_enricher(
        auth: AuthResult,
        params: dict[str, str],
    ) -> list[str]:
        features = await resolve_features(auth.identity)
        return [f"{opts.role_prefix}:{slug}" for slug in features]

    return entitlement_role_enricher
